### ISSUE TITLE
[promise-function-async] "allowedPromiseNames" misnomer.

### ISSUE THREAD

**Issue by sripberger** [typescript-eslint/typescript-eslint#1572]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/promise-function-async": [
        "error",
        { "allowedPromiseNames": [ "QueryBuilder" ] }
    ]
  }
}
```

```TS
import { Foo } from './models/foo';
import { QueryBuilder } from 'objection';

function foo(): QueryBuilder<Foo> {
    return Foo.query();
}
```

**Expected Result**

No errors.

**Actual Result**

```
error  Functions that return promises must be async  @typescript-eslint/promise-function-async
```

**Additional Info**

I'm looking to confugre the `promise-function-async` rule such that it allows non-async functions returning certain types, even if they are promises or promise-like. My use case is the Objection.js [QueryBuilder][1] which is a promise-like that executes the query you're building as soon as its `then` method is called. I need a function that returns it without executing it in order to do further manipulation, and adding `async` to this function will make the intended use of this function confusing.

Based on the name, it seems like the "allowPromiseNames" option should do this, but according to both [the docs][2] and implemented behavior shown above, it actually does the opposite. Instead of *allowing* the provided list of promise names, it considers the provided names promises even if they weren't before.

Note that in my case, this error is reported with or without the "allowedPromiseNames" option present, since QueryBuilder is a promise-like.

I'd like to add a true "allowPromiseNames" option, and rename the existing one to something else. Possibly "additionalPromiseNames." I could add the configuration under a different name, but I feel the presence existing one would be confusing.

Of course, this would be a breaking change, so I'd like to ensure this seems reasonable to you all before making a PR.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.19.0` |
| `@typescript-eslint/parser`        | `2.19.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `10.16.0` |
| `npm`                              | `6.13.7` |

[1]: https://vincit.github.io/objection.js/api/query-builder/
[2]: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/promise-function-async.md#options

**Comment by bradzacher** [comment]
Having a quick look at the library, it's actually got incorrect types. It says that `QueryBuilder` extends `Promise`, when it doesn't actually; it implements `Promise`, but isn't a Promise.

This is why the rule reports on those return types, because `Promise` is in the inheritance chain for `QueryBuilder`.

Looking even closer, I don't think it actually implements `Promise`, because it doesn't implement the ES2018 `finally` method, nor does it implement `[Symbol.toStringTag]: string`;  both of which create a type-safety hole.

----

I don't see any reason to rename the options and introduce a breaking change.

It may not be the clearest name, but I think its function is clearly spelled out in the docs, and you have to discover the options via the docs.

> `allowedPromiseNames` to indicate any ***extra names*** of classes or interfaces to be ***considered Promises*** when returned.

The problem is that you're reading it from your specific and rare use case of wanting to ignore a certain name. If you instead read it as "allow the rule to check these names as if they were promises", then the name makes sense.

I'd instead consider adding `ignorePromiseNames`, but I feel like this is a really far reaching and dangerous option, because it lacks context. I generally fight against those sorts of options for this reason.

What do I mean by lacking context? 
You have some functions which return `QueryBuilder`, with the intention that it's typed as such so that it can be consumed as a query builder.
Makes sense that this function isn't marked async because it's not doing any async work yet.

But what about a function which calls one of those query functions, does some work with the builder, and then returns the `QueryBuilder` with the intention that it's used as a promise?
Well because the ignore option is contextless, that function will also be ignored.
Sure this can be worked around by either calling the `execute` method, or explicitly annotating that function with a return value `: Promise<T>`, but then you have to explicitly remember to do those things.

```ts
// function shouldn't be async because it's intentionally returning a query builder - ignore option works as intended
function doQuery(...args): QueryBuilder {
  // ...
}

// should this function be async? The intention is that the caller should just await the value, but because of the ignore option it won't get flagged.
function returnPromise() {
  return doQuery('SELECT * FROM users;');
}
```

What I have found with rules pertaining to types and their names is that generally you don't want to _always_ ignore a thing, you want to instead ignore a thing _in certain cases_.
My recommendation is to generally use eslint disable comments, or eslint overrides configs for certain folders/filename patterns.

**Comment by sripberger** [comment]
So, I can understand how one might have arrived at the use of the word "allow" when naming this particular option, and it *can* make sense when viewed from the angle you've described.

In the context of eslint configuration, however, the word "allow" typically indicates relaxing of a rule in a certain scenario. `no-underscore-dangle` has "allowAfterSuper," "allowAfterThis," and even a generic "allow" option where you can explicitly list identifiers with dangling underscores that are allowed in all cases. There are many, many examples of this. I don't think it would be a stretch to call it idiomatic within the eslint community. The naming of this option goes against that idiom, hence my concern.  Especially if I intend to add an option that *would* follow the idiom.

As for your objections to such an option, I'd say that it isn't really possible to idiot-proof eslint, considering how configurable it already is. If you're worried about people misusing this option, that boat sort of already sailed in the sense that many organizations will probably just turn this rule off completely if it is getting in their way the way the way it's currently getting in mine.

Most organizations, to be honest, will probably never have it turned on to begin with since it isn't in the recommended set, and it produces unresolvable conflicts with `require-await`, which *is* in the recommended set. If someone is configuration this option at all, they're probably well-aware of the specifics of what they're allowing or disallowing.

In my case, I want every asynchronous operation to be marked with `async`, regardless of whether it uses an `await` or not. This is to help present consistent docs to vscode intellisense, and to prevent confusion about intent exactly along the lines of what you're describing.

I also require every async function to provide an explicit return type of `Promise<T>`, with T being whatever it resolves with. As such, I will *never* encounter a scenario where someone might return a QueryBuilder that they *intend* to be consumed as a promise without explicitly *calling* it a Promise. The intent is required to in the type signature itself.

The reason why I'd rather have control over relaxing this specifically for QueryBuilder is because I'd otherwise prefer to keep this rule in place. I don't want to disable it completely for entire files or directories, and I don't want developers to have to surround every single one of these functions with disabling comments. I just want to relax it in this very specific case.

That being said, if this really is a mistake on the part of Objection's maintainers, and QueryBuilders should not be getting identified as promises at all, this is all kind of moot for me since my specific case could be resolved by a bugfix in that library. I'll be looking into that as well, but seeing as a QueryBuilder *can* be `await`ed or `then`ed, it didn't really seem too unusual to me. I'm not too familiar with the criteria that will cause this rule to trigger. :\

**Comment by bradzacher** [comment]
> and it produces unresolvable conflicts with require-await

It shouldn't do. That's the point of our extension version; it should not false positive if you return an awaitable thing. If you have a case where they don't line up, please let me know.

> it isn't really possible to idiot-proof eslint

No, but it's possible to be thoughtful about our API design, and ensure changes are done for the right reasons.
I always push back on feature requests when I think there is another way to do things, and/or the change has a far-reaching effect.

ignore options are a bit of a pet peeve of mine, because whilst they're often done with good intentions, they end up creating a much larger hole than intended, and there's usually a better way to deal with the specific case.
A lot of time people also fire off issues and then forget them, or don't have reasons for wanting an ignore beyond "because I want it", so pushing back either lets me know that it's not a real request, or lets me discuss and figure out the actual intended use case. From there we often arrive at a solution which actually solves their problem. Only a few times has this resulted in an ignore option.

> if this really is a mistake on the part of Objection's maintainers

I would say that it is for the reasons that I mentioned above. For reference as well, `bluebird` is an npm package which creates its own promise implementation (i.e. just like Objection, its classes do not extend Promise, but look like Promises), and its types don't `extends Promise`, they `implements PromiseLike`:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/28952aa6e7ab947dbd7c40f90c5208287d20ab06/types/bluebird/index.d.ts#L43
Libraries like `bluebird` are actually the entire reason this option was added, because people use bluebird _instead of_ native promises.

Objection using `extends Promise` is completely understandable - it saves them copy pasting and maintaining the method signatures themselves, and _for the most part_ it won't cause issues. 

Buuuut there are some cases where it causes real issues:
- this case here, we're we're looking for things that are actually promises
- `instanceof Promise` checks:
```ts
declare class QueryBuilder<T = void> extends Promise<T> { }
declare class Bluebird<T = void> implements PromiseLike<T> {
    // ... snip - see repl link below ...
}

function test(prom: QueryBuilder | Bluebird) {
    if (prom instanceof Promise) {
        prom;
        // typeof prom === QueryBuilder ❌❌❌
        // THIS IS VERY, VERY WRONG
    } else {
        prom;
        // typeof prom === Bluebird ✅✅✅
    }
}

function acceptsPromise(prom: Promise<void>) { }
declare const extendsPromise: QueryBuilder;
declare const likePromise: Bluebird;

acceptsPromise(extendsPromise); // no errors ✅✅✅
acceptsPromise(likePromise);    // no errors ✅✅✅
```
[repl](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=29&pc=49#code/CYUwxgNghgTiAEkoGdnwIoFcQwJ4CFMBLCUGAHgBV4BeeANwHsjgA+eEADwBcQA7YGgAKMRgFsiyEFXYBveAF8AUKCRxE0VPHwRsAIyIxgVWg2Zt4RMQAcIIMf27DREqQBkiAa2mU5S+AHw3AAW-FQASiDImBDcAIymlAA08JSR0bEATKZ8IPQ4rAAU-oGljHwAZjEVJHbAAPwAXPCFhfRQuiDNlACUtOxpUTHx8AA+8CLikiAe3hFDsXGsfeOYAiA1ucBj8HwxEEklpQHlcABW4LwNza1wKOXNUHy4fTQD6cPZ45OuM14+Hyyyx2a1AmxA23GewgByO8B6zR0+kMxkGGRG4zRn1YAG4lHCwFBuGBgvN0Tk8gVCqcQBcwFcmi1CndkA94E8Xv1UoDuDsftNZgCFtxgat1uDIbt9gjtJ0DEYTJiebj8aUANoAZVwYj0jAgADpuIwNdwYEQ+ABzShQC0AXWayFN5oteOUSiqfHpRHKQSi3EK1hczSwOAIxFIOB2SJA8uAfVkcKIFRagfElj4jqeYBAjGT-Kk8bhpVTYjxx0CAHoK0FcNYc8mS7QaHQQ3hCCQyPBADLkPaLlerlAAEgBJDXwUfwABqAFFwgBNFIz+fwADq4QA8gA5ADicIUHAgUngCfLARLZdPVZrddz8Ebzbo0dj8EAoORvvdKN0er0+qBgbPWE4+YgAGQYTC40zkEwLDAvIyiqNA6hgOUjocDw-CCMBwbYG24ZkHiCGwAgyEZryED-FhsrIkYeJKH+AFARBUiFFwvACMgwE9Di8BXnwjAcDAogwGgb4vnR-4gIBHFMSB5HeJx3H9rs-E4EJIlvkAA)

**Comment by sripberger** [comment]
That all sounds completely reasonable to me. I will say, though, that I don't really intend for this to be a fire and forget kind of issue. This is a real problem for me that I intend to solve one way or another. If that involves just turning off the rule for a subset of files, I'll do that, though the sticker in me wants a more correct approach.

You do make a good point about this possibly being Objection's problem. I may see if I can change Objection to avoid triggering the rule, though I'm a lot less famliar with that codebase than I am with tweaking eslint rules. In the meantime I will probably just disable the rule wherever needed. I'll consider the ignore option a last resort, as you've said it should be.

As for `require-await`, I was referring to the original eslint rule. I originally created my configs before the extension version of this rule existed, and I haven't been keeping up with all the new features. It's super cool to see that extension version, though. Thanks for pointing it out. :+1:

**Comment by bradzacher** [comment]
>  I may see if I can change Objection to avoid triggering the rule, though I'm a lot less famliar with that codebase

At least from my quick look whilst researching my first comment, it _looks like_ it should be a relatively simple change to just this file:
https://github.com/Vincit/objection.js/blob/5b76a4c1e0dee70649d32b8a59383021bb5d5927/typings/objection/index.d.ts#L798

Their codebase is otherwise in pure JS, so I think this is just a problem with their hand-crafted types.
Should be simple enough to add signatures for `then` and `catch`, and switch `extends Promise<R>` to `implements PromiseLike<R>`
